packages/playground/data-liberation/src/import/WP_Topological_Sorter.php

```diff
@@ -0,0 +1,406 @@
+<?php
+
+/**
+ * The topological sorter class. We create a custom table that contains the WXR
+ * IDs and the mapped IDs. Everytime an entity is processed, we add it to the
+ * table. The first time we process an entity, it is mapped to the original ID
+ * and no mapped ID. From the second time, it is mapped to the mapped ID.
+ *
+ * When the WP_Entity_Importer or similar class read raw data from the source
+ * stream that is used to map the original IDs to the mapped IDs.
+ *
+ * The first STAGE_TOPOLOGICAL_SORT stage do save all the entities with no
+ * mapped IDs. So during the STAGE_IMPORT_ENTITIES step the WP_Entity_Importer
+ * or similar class read already inserted data and save them. From that moment
+ * all the entities have the IDs created using wp_insert_post(),
+ * wp_insert_comment(), wp_insert_term(), wp_insert_comment_meta(),
+ * wp_insert_post_meta() and wp_insert_term_meta() calls.
+ */
+class WP_Topological_Sorter {
+
+	/**
+	 * The base name of the table.
+	 */
+	const TABLE_NAME = 'data_liberation_map';
+
+	/**
+	 * The option name for the database version.
+	 */
+	const OPTION_NAME = 'data_liberation_db_version';
+
+	/**
+	 * The current database version, to be used with dbDelta.
+	 */
+	const DB_VERSION = 1;
+
+	/**
+	 * The current session ID.
+	 */
+	protected $current_session = null;
+
+	/**
+	 * The current item being processed.
+	 */
+	protected $current_item = 0;
+
+	/**
+	 * The entity types saved in the database.
+	 */
+	const ENTITY_TYPES = array(
+		'comment'      => 1,
+		'comment_meta' => 2,
+		'post'         => 3,
+		'post_meta'    => 4,
+		'term'         => 5,
+		'term_meta'    => 6,
+	);
+
+	/**
+	 * The name of the field where the ID is saved.
+	 */
+	const ENTITY_TYPES_ID = array(
+		'comment'      => 'comment_id',
+		'comment_meta' => 'meta_key',
+		'post'         => 'post_id',
+		'post_meta'    => 'meta_key',
+		'term'         => 'term_id',
+		'term_meta'    => 'meta_key',
+	);
+
+	/**
+	 * Set the current session ID.
+	 */
+	public function __construct( $options = array() ) {
+		if ( array_key_exists( 'session_id', $options ) ) {
+			$this->set_session( $options['session_id'] );
+		} else {
+			$active_session = WP_Import_Session::get_active();
```

Comment:
What if we're running without a WP_Import_Session?

Reply:
Ok.